src/AzureAuth/Commands/Ado/CommandToken.cs

```diff
@@ -3,59 +3,127 @@
+        private string PromptHint { get; set; }
 
         /// <summary>
-        /// Gets or sets the prompt hint.
+        /// Format a PAT based on <paramref name="output"/>.
         /// </summary>
-        [Option(CommandAad.PromptHintOption, CommandAad.PromptHintHelpText, CommandOptionType.SingleValue)]
-        public string PromptHint { get; set; }
+        /// <param name="value">The PAT value.</param>
+        /// <param name="output">The output mode.</param>
+        /// <param name="scheme">The <see cref="Authorization"/> scheme to use.</param>
+        /// <returns>The formatted PAT value ready for printing.</returns>
+        public static string FormatToken(string value, OutputMode output, Authorization scheme) => output switch
+        {
+            OutputMode.Token => value,
+            OutputMode.Header => value.AsHeader(scheme),
+            OutputMode.HeaderValue => value.AsHeaderValue(scheme),
+            _ => throw new ArgumentOutOfRangeException(nameof(scheme)),
+        };
 
         /// <summary>
         /// Executes the command and returns a status code indicating the success or failure of the execution.
         /// </summary>
         /// <param name="logger">The <see cref="ILogger{T}"/> instance that is used for logging.</param>
+        /// <param name="env">An <see cref="IEnv"/> to use.</param>
         /// <param name="eventData">Lasso injected command event data.</param>
         /// <returns>An integer status code. 0 for success and non-zero for failure.</returns>
-        public int OnExecute(ILogger<CommandToken> logger, CommandExecuteEventData eventData)
+        public int OnExecute(ILogger<CommandToken> logger, IEnv env, CommandExecuteEventData eventData)
         {
-            logger.LogInformation("coming soon");
-            return 0;
+            // First attempt using a PAT.
+            var pat = PatFromEnv.Get(env);
+            if (pat.Exists)
+            {
+                logger.LogDebug($"Using PAT from env var {pat.EnvVarSource}");
+                logger.LogInformation(FormatToken(pat.Value, this.Output, Authorization.Basic));
+                return 0;
+            }
+
+            // If no PAT then use AAD AT.
+            var authResult = AzureAuth.Ado.TokenFetcher.AccessToken(
+                logger: logger,
+                mode: this.AuthModes.Combine().PreventInteractionIfNeeded(env),
+                domain: this.Domain,
+                prompt: AzureAuth.PromptHint.Prefixed(this.PromptHint),
+                timeout: TimeSpan.FromMinutes(this.Timeout));
+
+            var authflow = authResult.Success;
+            if (authflow != null)
+            {
+                logger.LogDebug($"Acquired AAD AT via {authflow.AuthFlowName} in {authflow.Duration.TotalSeconds:0.00} sec");
+                logger.LogInformation(FormatToken(authflow.TokenResult.Token, this.Output, Authorization.Bearer));

```

Comment:
If `authflow` is not null, we can completely ensure we have received an access token?

Reply:
Yes, because `authflow` is declared just above as the `authResult.Success` - which is either a succesful AuthFlowResult or null.